RFC: 0107 update document with graduation tasks (#133)
aws/aws-cdk-rfcs: text/0107-construct-library-module-lifecycle.md

- Stage 2 - Developer Preview
- Stage 3 - GA

Each stage in this document describes 3 distinct sets of activities:

> * **Entry activities:** Actions that need to be taken before the module enters this stage. These can either be automatic, or done by the team. They will usually be performed once a module is deemed for stage maturation, and all its exit criteria are met.

Reviewer: This is weird. One of our linter rules is that line length should not exceed 150 - https://github.com/aws/aws-cdk-rfcs/blob/master/tools/linters/markdownlint.json#L6-L8.

Wonder why this is not honoured. While I investigate this, can you fix this up? And in other places in the doc.

Reviewer: In the mean time I fix this, please run the commands on here https://github.com/aws/aws-cdk-rfcs/blob/master/.github/workflows/pr-lint.yml and fix all violations.

Reviewer: I found the issue and fixed it up here - https://github.com/aws/aws-cdk-rfcs/pull/136

Sorry :(

Author: 👍

> * **Entry activities:** Actions that need to be taken before the module enters this stage. These can either be automatic, or done by the team. They will usually be performed once a module is deemed for stage maturation, and all its exit criteria are met.

Reviewer: "They will usually be performed once a module is deemed for stage maturation"

"They will be performed as soon as the module moves into the new stage"

Author: 👍

> * **Ongoing activities:** These are activities performed during the time a module lies statically in the given stage (e.g bug fixes, new features...)

Reviewer: "lies statically" implies nothing is changing.

How about "These are the activities that are primarily performed when a module is in that specific stage"?

Author: 👍

> * **Exit criteria:** A module can progress to the next stage only if its exit criteria are met. Note that this is required, but not sufficient. A module matures to the next stage only if its exit criteria are met, **and** was prioritized over other modules using the [prioritization framework](#Construct-prioritization-framework).

Reviewer: How about starting off with "When these criteria are all met, a module progresses into the next stage. Note that, while all of the criteria here are necessary, they may not be sufficient for a module to progress."?

> A module matures to the next stage only if its exit criteria are met, **and** was prioritized over other modules using the [prioritization framework](#Construct-prioritization-framework).

I'll probably drop this part. This is true for everything across this doc.

Further, if this is the only reason you've called 'necessary but not sufficient', probably better to drop that out as well.

Author: > I'll probably drop this part. This is true for everything across this doc

Yes, but this is a global section that relates to the entire doc as well. I think its fitting and actually this reasoning isn't stated anywhere else - is it? It also combines well with your comment here: https://github.com/aws/aws-cdk-rfcs/pull/133#discussion_r387101832

Reviewer 2: I think this is okay to keep and is in the section that applies to the entire document. However, I think that we all agreed that they are prioritization tenets rather than a framework.

Author: Yeah, but lets push that change in a different PR. This isn't the only reference to it as a framework in the doc. I think its ok to leave it as is for now.

## Stage 0 - CFN Resources

All construct library modules start in stage 0 when they are auto-generated from the CloudFormation resource
specification. The goal of stage 0 is to make new CloudFormation resources/properties available to CDK customers as soon
as possible, and create tracking artifacts that allow us to capture the data required to make prioritization decisions
to add L2 resources in the future.

### Entry activities

>   > Many actions here are actually performed automatically during our weekly release process.

Reviewer: ```suggestion
  > The following actions are performed automatically during the weekly CDK release process:
```

Author: 👍

  > - Generation of the module L1 code when CFN introduces a new spec.
  > - Generation of the initial module README.
  > - Mark module as "experimental" in `package.json`.
  >
  > The remaining activities described here need to be done manually.

>   - **Create GitHub tracking issue** - Create a GitHub tracking issue for the module. The primary role of the tracking

Reviewer: Minor: indenting. Now that the stage names are headers, you can drop the 2 space indent and align all of them right at the margin.

Author: Cool

as possible, and create tracking artifacts that allow us to capture the data required to make prioritization decisions
to add L2 resources in the future.

> - **Entry activities:**

Author: 👍

>   - **Automatically publish package** - A package for the module is autogenerated and published in the aws-cdk GitHub

Author: 👍

>   - **Automatically publish package** - A package for the module is autogenerated and published in the aws-cdk GitHub

Author: Decided to put this as the first paragraph inside the section because it feels more appropriate. Let me know if its clear enough.

  - **Validate API reference** - Validate the module is listed in the AWS Construct Library API reference and is
    rendered correctly

> - **Ongoing activities:**

Author: Your'e right, I actually forgot about this at the last minute. Thanks 👍

  - **Validate API reference** - Validate the module is listed in the AWS Construct Library API reference and is
    rendered correctly

### Ongoing activities
  - **Observe** - Watch KPIs to prioritize the construct library module's graduation:
    - pull requests that add the first high-level constructs are the strongest signal to graduate a construct library
      module from stage 0 to stage 1 (experimental) and are prioritized first (this is a manual activity)
    - customer "+1s" on a construct library module's tracking issue exceeding 100 positive reactions, are prioritized
      next
    - construct library modules that accumulate 10+ feature requests, are prioritized next
### Exit Criteria
  - when the first community PR for a module is merged, OR

>   - when selected by the CDK team using the [prioritization framework](#Construct-prioritization-framework)

Reviewer: Sorry this caught my eye. Can we just say, "when the first PR for this module that introduces constructs is merged"

Author: Agreed.

visibility. See [Appendix A](#Appendix-A:-GitHub-Artifacts) for an illustration of how these GitHub artifacts are used
and related to each other.

### Entry activities

- **Identify key stakeholders** - Opportunistically recruit stakeholders outside the CDK team to participate in API
  reviews and usability studies. Potential stakeholders include:
  - service's product manager,
  - service's developer advocates (if any),

>   - service's software development engineer(s) (designated by service's leadership,

Reviewer 2: close parenthesis

Author: 👍

- **Clear awslint exclusions** - Review `awslint` exclusions and make the necessary code changes to remove what we can.

- **Create Graduation Issue** - Create a GitHub issue titled `Graduate @aws-cdk/{module} to stable`.
This issue communicates our intent to graduate this module. It will also serve as an asynchronous channel for anyone who wants to provide feedback.
Update the module tracking issue with a link to this issue so that the community can be notified.

  > External Stakeholders and major community contributors should be contacted directly.

- **Conduct API review**

  Champion to conduct a final API review.

  > The champion is expected to be relatively versed in the library.
  > If this requires some effort to achieve, the review should be scheduled with this in mind.

  The goal of this review is to identify critical API and Security issues.
  It is meant to be as frictionless and slim as possible. Attendees are:

  - Module owners (Required)
  - Graduation Champion (Required)
  - Other CDK Team Members (Optional)
  - External Stakeholders (Optional)
  - Community Contributors (Optional)

>   The review will be held on chime and the invite should be sent no less than a week in advance.

Reviewer 2: Chime is a proper noun with a capital "C"

Author: 👍

      "Working on it" column
    - otherwise, place the module’s tracking issue on the CDK roadmap in the “Working on it“ column

  - **Update Module Stability Label** - In the module's `package.json`, set "Stability": "Experimental" to indicate the

>     high-level constructs are under active development and subject to breaking changes.

Reviewer: We already set this flag in the package.json as part of the package when it's first created (at stage-0).

This should be merged with the same paragraph mentinoned in the comment against the 'Automatically publish package' item.

Author: Good catch.

The goal of developer preview is to deliver a release candidate with a stable API to conduct user acceptance testing and
accumulate sufficient usage to declare a module is ready for general availability. We will only make breaking changes to
developer preview modules when required to address unforeseen customer use cases or issues. Since the potential for

> breaking changes still remains while in Developer Preview, the package name will keep the“experimental” stability during

Reviewer: Minor: space between `the` and `"experimental"`. 

Minor: use standard double quotes (U+022). It looks like this is using U+201C and U+201D.

  - **Clear awslint exclusions** - Review `awslint` exclusions and make the necessary code changes to remove what we can.

  - **Create Graduation Issue** - Create a GitHub issue titled `Graduate @aws-cdk/{module} to stable`. This issue communicates our intent to graduate this module. It will also serve as an asynchronous channel for anyone who wants to provide feedback. Update the module tracking issue with a link to this issue so that the community can be notified.

    > External Stakeholders and major community contributors should be contacted directly.

  - **Conduct API review**

    Champion to conduct a final API review.

    > The champion is expected to be relatively versed in the library. If this requires some effort to achieve, the review should be scheduled with this in mind.

    The goal of this review is to identify critical API and Security issues. It is meant to be as frictionless and slim as possible. Attendees are:

    - Module owners (Required)
    - Graduation Champion (Required)
    - Other CDK Team Members (Optional)
    - External Stakeholders (Optional)
    - Community Contributors (Optional)

>     The review will be held on chime and the invite should be sent no less than a week in advance. This is to give the attendees a chance to prepare if they so desire. Include the agenda in the invite:

Reviewer: Nice! Love all the detail in these steps here. I'm hoping, one day in the near future, this whole doc would be as detailed.

    - Champion to introduce the people and re-iterate the goal of the review.
    - 10-15 minutes individual reading of the module README.
    - Participants raise concerns.
    - Champion will do a quick overview of external APIs not covered by the README.
    - Participants raise concerns.
    - Champion will address any asynchronous feedback provided in the graduation issue.
    - Participants raise concerns.
    - Champion will address any high traction open `guidance` issues.
    - Participants raise concerns.
    - Champion will summarize and recap all action items.

>     - **Participants decide if the module stays in *Developer Preview* or if its ready to move to *Stable*. In addition, decide which action items need to be completed before graduation.**

Reviewer: Wonder if we should make the call here *necessarily*. 

Perhaps the library can be announced as graduating (in the meeting) if no serious issue is raised. If there were any serious issues, the decision can be asynchronous (over email); just like how we did for custom-resource. Folks may want time to research or think it over?

Author: Decided to put this in a quote comment as an additional valid decision. Let me know what you think.
